curs/curs-16-17-optimizari/README.md

```diff
@@ -0,0 +1,217 @@
+# Optimizari
+
+## Introducere
+
+Optimizarea reprezinta procesul prin care randamentul unei aplicatii
+(sau a unui sistem) este maximizat in functie de un anumit criteriu.
+Acest criteriu poate fi timpul, spatiul sau energia consumata.
+
+Avand in vedere complexitatea sistemelor din ziua de azi - o aplicatie
+nu ruleaza niciodata in vid, ci poate rula direct deasupra unui sistem
+fizic (baremetal), ori deasupra unui sistem de operare care la randul
+sau poate rula in cadrul unei masini virtuale sau direct deasupra unui
+sistem fizic etc. -, procesul de optimizare este unul complicat, multinivel
+pentru care nu exista o cale batatorita ce poate fi urmata. In aceste
+conditii, optimizarea unui program necesita profilarea extensiva a acestuia
+pentru a se descoperi nivelurile unde au loc pierderi de performanta si
+cauzele acestora.
+
+Din punct de vedere al programatorului optimizarea poate avea loc la 3
+niveluri diferite:
+
+1. Optimizari la nivel de algoritm. Acest pas presupune identificarea
+algoritmului optim pentru problema de fata si este independent de
+ecosistemul in care ruleaza aplicatia (ceea ce am invatat la "Analiza
+Algoritmilor").
+
+2. Optimizari la nivel de implementare. In acest punct este necesar ca
+programatorul sa cunoasca detalii despre arhitectura sistemului pe care
+ruleaza si prin indicii sau rescrieri de cod sa ajute compilatorul sa
+genereze un cod mai eficient.
+
+3. Optimizari la nivel de asamblare. Acest pas presupune intelegerea
+codului assembly rezultat si identificarea situatiilor in care compilatorul
+nu a generat codul optim. In acest caz, programatorul va trebui sa
+substituie sau sa adauge cod (inline) assembly in codul sursa de nivel
+inalt.
+
+In cadrul acestui curs ne vom axa pe optimizarile de timp la nivel de
+asamblare pentru a vedea cum cunoasterea arhitecturii sistemului de
+calcul si a limbajului de asamblare ne poate ajuta sa optimizam
+programele. Astfel, cursul va aborda urmatoarele subiecte:
+ - vom enumera care sunt metodele de masurare/profilare disponibile
+ - vom analiza o serie de optimizari pe care le are la dispozitie compilatorul
+ - vom studia o serie de instructiuni speciale ale procesorului pe care
+   compilatorul nu le foloseste prea des, dar care au potential mare de a
+   creste performanta
+
+## Profilare
+
+Programele din ziua de azi sunt din ce in ce mai mari (de ordinul milioanelor
+de linii de cod). In acest context este important sa avem o metoda usoara
+de a identifica care sunt portiunile din program care sunt executate cel
+mai des (hotpaths). Odata identificate aceste portiuni, este necesar
+sa intelegem anatomia acestora pentru a putea gasi metode de optimizare.
+Profilarea unei aplicatii reprezinta procesul prin care putem realiza
+acest lucru.
+
+Astfel, in procesul de profilare vom utiliza una sau mai multe unelte,
+in functie de scop.
```

Comment:
De spus că există două tipuri mari de profilare:
* cu instrumentare (gprof)
* cu sampling (perf)

Reply:
Nu as intra in asemenea detalii. As vrea doar sa inteleaga care sunt scenariile de masurare (de la granularitate foarte mare pana la granularitate foarte fina) si faptul ca exista instrumente pentru fiecare caz. Cum functioneaza unealta intern este mai putin relevant.